tutorials/dynamic-web-forms-with-validation-in-java/content.adoc

```diff
@@ -0,0 +1,295 @@
+    ComboBox<String> snackTypeSelect = new ComboBox<>("Type", Collections.emptyList());
+    ComboBox<String> snackSelect = new ComboBox<>("Snack", Collections.emptyList());
+    Button orderButton = new Button("Order");
+    Div errorsLayout = new Div();
+
+    // Configure UI components
+    orderButton.setThemeName("primary");
+
+    // Wrap components in layouts <3>
+    HorizontalLayout formLayout = new HorizontalLayout(nameField, quantityField, snackTypeSelect, snackSelect, orderButton);
+    Div wrapperLayout = new Div(formLayout, errorsLayout);
+    formLayout.setDefaultVerticalComponentAlignment(Alignment.BASELINE);
+    wrapperLayout.setWidth("100%");
+
+    return wrapperLayout;
+  }
+}
+----
+<1> On the top level, we have three components: the heading, the form, and the grid. The form creation is complicated, so we move that into a separate method.
+<2> Create UI components for each of the properties in `SnackOrder`.
+<3> Define how the components should get laid out. `HorizontalLayout` displays all the fields next to each other, and the `Div` wrapping the form and error messages puts them on top of each other. 
+
+
+Run the application through maven: 
+
+[source]
+mvn package jetty:run
+
+Navigate to http://localhost:8080, and you should now see the following:
+
+.Form and data grid UI components.
+image::java-form-and-grid-ui-components.png[Form and data grid UI components.]
+
+== Dynamically populate a select based on a form value
+We need some options for snacks to order. Add the following Map of snack options at the beginning of the `buildForm` method:
+
+.`*MainView.java*`
+[source,java]
+----
+  private Component buildForm() {
+
+    Map<String, List<String>> snacks = new HashMap<>();
+    snacks.put("Fruits", Arrays.asList("Banana", "Apple", "Orange", "Avocado"));
+    snacks.put("Candy", Arrays.asList("Chocolate bar", "Gummy bears", "Granola bar"));
+    snacks.put("Drinks", Arrays.asList("Soda", "Water", "Coffee", "Tea"));
+
+    // remainder omitted ...
+  }
+----
+
+
+Then update the snack type select to display the different types of snacks. 
+
+[source, diff]
+----
+- ComboBox<String> snackTypeSelect = new ComboBox<>("Type", Collections.emptyList());
++ ComboBox<String> snackTypeSelect = new ComboBox<>("Type", snacks.keySet());
+----
+
+Next, disable the snack selection `ComboBox` initially, and add a listener on the snack type select that can be used to enable it with the correct options based on the type selection. 
+
+.`*MainView.java*`
+[source,java]
+----
+  private Component buildForm() {
+    // Field creation
+
+    // Only enable snack selection after a type has been selected.
+    // Populate the snack alternatives based on the type.
+    snackSelect.setEnabled(false);
+    snackTypeSelect.addValueChangeListener(e -> {
+      String type = e.getValue();
+      snackSelect.setEnabled(type != null && !type.isEmpty());
+      if (type != null && !type.isEmpty()) {
```

Comment:
This example is pure gold! Just saw equivalent implemented with FilterableCustomDataProviderExtensionWithMultipleParametersFactoryBean (or something) in our actual docs and it made me feel sick.